modules/core/src/smithy4s/Validator.scala

```diff
@@ -0,0 +1,64 @@
+/*
+ *  Copyright 2021-2024 Disney Streaming
+ *
+ *  Licensed under the Tomorrow Open Source Technology License, Version 1.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *     https://disneystreaming.github.io/TOST-1.0.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package smithy4s
+
+sealed trait Validator[A, B] {
+  def validate(value: A): Either[String, B]
+}
+
+object Validator {
+
+  def of[A, B]: PartialValidatorBuilder[A, B] =
+    new PartialValidatorBuilder[A, B]
+
+  private[smithy4s] class PartialValidatorBuilder[A, B] {
+    def validating[C](constraint: C)(implicit
+        bijection: Bijection[A, B],
+        ev: RefinementProvider.Simple[C, A]
+    ): ValidatorBuilder[A, B] =
+      new ValidatorBuilder[A, B](ev.make(constraint), List.empty, bijection)
+  }
+
+  private[smithy4s] class ValidatorBuilder[A, B](
+      first: Refinement.Aux[_, A, A],
+      other: List[Refinement.Aux[_, A, A]],
+      bijection: Bijection[A, B]
+  ) {
+
+    def validating[C](constraint: C)(implicit
+        ev: RefinementProvider.Simple[C, A]
+    ): ValidatorBuilder[A, B] =
+      new ValidatorBuilder[A, B](first, other :+ ev.make(constraint), bijection)
+
+    def build: Validator[A, B] = new ComposedValidator(first, other, bijection)
+  }
+
+  private final class ComposedValidator[A, B](
```

Comment:
this class and `ValidatorBuilder` have exactly the same data structure. There's no reason to separate the `Validator` with its `Builder` at all then, as you could very well just have a `def alsoValidate[C](constraint: C)(implicit ...) : ValidatorBuilder[A, B]` directly on the `Validator` interface.

Reply:
Reworked this a bit and got rid of the multiple "builders" and added, as suggested, a new `alsoValidate` method.